enigma-principal/app/src/epoch_u/epoch_provider.rs

```diff
@@ -56,146 +54,245 @@ impl EpochProvider {
-                };
-                epoch_state
+                let mut data: Vec<EpochState> = Deserialize::deserialize(&mut des).unwrap_or_default();
+                println!("Found EpochState list: {:?}", data);
+                if cap < data.len() {
+                    return Err(EpochStateIOErr { message: format!("The EpochState entries exceed the cap: {}", cap) }.into());
+                }
+                let mut capped_data = Vec::with_capacity(cap);
+                capped_data.append(&mut data);
+                capped_data
             }
             Err(_) => {
                 println!("No existing epoch state, starting with block 0");
-                None
+                vec![]
             }
         };
         Ok(epoch_state)
     }
 
     #[logfn(DEBUG)]
-    fn write_epoch_state(epoch_state: Option<EpochState>) -> Result<(), Error> {
-        let path = Self::get_state_file_path()?;
-        match epoch_state {
-            Some(epoch_state) => {
-                let mut file = File::create(path)?;
-                let mut buf = Vec::new();
-                epoch_state.serialize(&mut Serializer::new(&mut buf))?;
-                file.write_all(&buf)?;
-            }
-            None => {
-                match fs::remove_file(path) {
-                    Ok(res) => println!("Epoch state file removed: {:?}", res),
-                    Err(_err) => eprintln!("No epoch state file to remove"),
-                }
-            }
-        };
+    fn write_to_file(epoch_state: Vec<EpochState>) -> Result<(), Error> {
+        let path = Self::get_file_path()?;
+        if epoch_state.is_empty() {
+            return Ok(fs::remove_file(path).unwrap_or_else(|_| println!("No epoch state file to remove")));
+        }
+        let mut file = File::create(path)?;
+        let mut buf = Vec::new();
+        epoch_state.serialize(&mut Serializer::new(&mut buf))?;
+        file.write_all(&buf)?;
         Ok(())
     }
 
-    #[logfn(DEBUG)]
-    fn parse_worker_parameterized(&self, receipt: &TransactionReceipt) -> Result<Log, Error> {
-        let log = receipt.logs[0].clone();
-        let raw_log = RawLog { topics: log.topics, data: log.data.0 };
-        let event = WorkersParameterizedEvent::new();
-        let result = match event.0.parse_log(raw_log) {
-            Ok(result) => result,
-            Err(err) => return Err(Web3Error {
-                message: format!("Unable to parse {} event: {:?}", WORKER_PARAMETERIZED_EVENT, err),
-            }.into()),
-        };
-        println!("Parsed the {} event: {:?}", WORKER_PARAMETERIZED_EVENT, result);
-        Ok(result)
-    }
-
-    /// Returns the `EpochState` stored in memory
-    pub fn get_state(&self) -> Result<EpochState, Error> {
-        let guard = match self.epoch_state.try_lock() {
+    /// Lock the `EpochState` list `Mutex`, or wait and retry
+    pub fn lock_guard_or_wait(&self) -> Result<MutexGuard<Vec<EpochState>>, Error> {
+        // TODO: retry if locked
```

Comment:
why not just `lock()` instead of `try_lock()`? then you wait for it to unlock.
otherwise this will result the same as here: https://github.com/enigmampc/enigma-core/issues/140

Reply:
and you don't need a special function for it. locking a mutex automatically wait for it to be available. https://doc.rust-lang.org/std/sync/struct.Mutex.html https://doc.rust-lang.org/std/sync/type.TryLockResult.html

Reply:
Got it. I misunderstood the purpose to `try_lock`. I assumed that it was the same as `lock` with handled exceptions. I was surprised that Mutexes in Rust didn't block and wait. I understand the distinction now after reading the doc more carefully. I'll submit this change in a separate PR.